### ISSUE TITLE
No error raised when using GQLRequestMutation type

### ISSUE THREAD

**Issue by shahzadlone** [sourcenetwork/defradb#2410]
There are cases where `GQLRequestMutation` type does not raise error when other mutation types do.

So to get around this in `UpdateDoc` test case we introduced a new prop called `NoGQLError` boolean which will pass the test even if no error is raised during action handler func `updateDoc` by adding this:

```go
if action.NoGQLError && mutationType == GQLRequestMutationType && !expectedErrorRaised {
	return	
}
```

When resolving this, make sure all error types that are returned from each mutation type is consistent.

Also likely can add another prop for explicit use of `DocID` instead of only allowing to use the current `DocID` which is an index. This would catch this case of inconsistent error for other more common places.

**Comment by AndrewSisley** [comment]
> which will pass the test

Please do not make the test pass, instead skip it.

**Comment by AndrewSisley** [comment]
Although, come to think about it I dislike this boolean.  We already had/have a means for handling this - the `TestCase` can have a property that says it only supports certain mutation types.  If it doesn't exist at the moment, you can find the old code in the git history and use that maybe, or just rewrite it.

The test should not look valid (and passing) and create a false sense of security.

**Comment by shahzadlone** [comment]
> Please do not make the test pass, instead skip it.

Why should we skip it, it needs to be asserted as it is asserting a correct behavior. It's just that the correct behavior in GQL case does not output a visible error. 


> Although, come to think about it I dislike this `NoGQLError` boolean. We already had/have a means for handling this - the `TestCase` can have a property that says it only supports certain mutation types. If it doesn't exist at the moment, you can find the old code in the git history and use that maybe, or just rewrite it.
> 
> The test should not look valid (and passing) and create a false sense of security.
> 
> EDIT: I checked, this still exists - see `TestCase.SupportedMutationTypes`, and just use that wherever you were thinking of using the boolean.


The test should still pass and assert even for GQL client, it's not a false sense of security, just that an extra error is not popping up, the behavior should still be asserted and caught if incase it's buggy. One additional thing I can do in the:

```
if action.NoGQLError && mutationType == GQLRequestMutationType && !expectedErrorRaised {
	return	
}
```

is to add an other condition where this only happens for that specific error string that we want to still pass for even if that error sting is not raised.

**Comment by AndrewSisley** [comment]
Lets talk about this in the standup.  I very much want all our clients/mutation types to work in the same way, it makes things a lot easier for us and the users IMO.  If we want them to behave the same way, then this missing error is a bug, if not, I still think a highly output specific `NoGQLError` is a poor solution to the problem you are trying to solve.

**Comment by shahzadlone** [comment]
> I very much want all our clients/mutation types to work in the same way

Same here, however, that seems to be an existing bug, GQL request on a document that does not exist will not return an error.

I am guessing the planner is the one that suppresses the error.

**Comment by AndrewSisley** [comment]
If it is a bug, the test should not pass, and it very definitely should not look like it passes.  This use case is exactly why `TestCase.SupportedMutationTypes` exists.

**Comment by shahzadlone** [comment]
> If it is a bug, the test should not pass, and it very definitely should not look like it passes. This use case is exactly why `TestCase.SupportedMutationTypes` exists.

I am not talking about my ACP test in my previous test. I am talking about the current state of non-acp behaviour.

**Comment by shahzadlone** [comment]
The action it self does not perform the mutation (which is correct behavior even for GQL request).

And the test action before and after that action assert the correct behavior is happening, if we skip all of these assertions will skip too.

The only incorrect thing in that test is that for GQL the error doesn't explicitly raise.

**Comment by AndrewSisley** [comment]
I am failing to see the problem.

We have some test actions to perform.  Some mutation types result in an error, some do not.  Write two sets of tests using `TestCase.SupportedMutationTypes`, one expecting the error, and one set without the error.

**Comment by shahzadlone** [comment]
> I am failing to see the problem.
> 
> We have some test actions to perform. Some mutation types result in an error, some do not. Write two sets of tests using `TestCase.SupportedMutationTypes`, one expecting the error, and one set without the error.

That suggestion makes sense. I thought you were saying to skip that test completely haha. I am happy with that route.

**Comment by shahzadlone** [comment]
Do we want to keep this ticket open for making our clients eventually behave the same in future?

**Comment by AndrewSisley** [comment]
> Do we want to keep this ticket open for making our clients eventually behave the same in future?

Yes please, and please tweak the description, and link to this item via comments above any relevant tests that you are writing (e.g. below), and also note the existence of those tests in this issue's description (if there are a lot, just suggest a Ctrl+f for the link or something).
```
// This test documents unwanted behaviour, for details please see:
// https://github.com/sourcenetwork/defradb/issues/2410
func TestFoo(t *testing.T) {
```
